### PR TITLE
LibCore+groupadd: Core::Group abstraction for group management + groupadd port to LibMain

### PR DESCRIPTION
**Pull request by kennethmyhra**
This PR introduces a an abstraction which currently can be used to add POSIX group database entries to `/etc/group`, as well as represent a POSIX group database entry. In the future it can be expanded to help us fetch and remove group database entries.

The motivation for creating the Core::Group abstraction was to hide the hybrid world and clunkiness of mixing C and C++ APIs when working with the POSIX Group Database. I.e: putgrent(), fopen(), struct group, etc.

####  LibCore: Add Core::Group abstraction for group management
This adds the Core::Group C++ abstraction to ease interaction with the group entry database, as well as represent the Group entry.
Core::Group abstraction currently contains the following functionality:
- Add a group entry - 'Core::Group::add_group()'

#### groupadd: Port to LibMain and use the new Core::Group abstraction :^)

### CODE REVIEW

File: AK/String.h [hunk 1]
```diff
@@ -160,6 +160,7 @@ class String {
 
     [[nodiscard]] bool is_null() const { return !m_impl; }
     [[nodiscard]] ALWAYS_INLINE bool is_empty() const { return length() == 0; }
+    [[nodiscard]] ALWAYS_INLINE bool is_null_or_empty() const { return !m_impl || length() == 0; }
```
Comment (awesomekling) [inline AK/String.h:163]: This is equivalent to `String::Is_empty()` (since `length()` null-checks `m_impl`)

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/Group.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    return TRY(from_libc_group(group.value()));
+}
+
+#ifndef AK_OS_MACOS
+ErrorOr<void> Group::add_group(Group& group)
+{
+    if (group.group_name().is_null_or_empty())
+        return Error::from_string_literal("Group::group_name() is required.");
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:37]: That's kind of a weird error message. Wouldn't `"Group name must not be empty."` / `"Group name is empty."` be better?

File: Userland/Libraries/LibCore/Group.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    , m_group_id(group_id)
+    , m_group_members(move(group_members))
+{
+}
+
+ErrorOr<struct group> Group::to_libc_group()
+{
+    struct group gr;
+    ::memset(&gr, 0, sizeof(struct group));
+    gr.gr_name = const_cast<char*>(group_name().characters());
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:90]: When `*this` is destroyed, `this->m_group_name` will be freed. This is a UAF.

File: Userland/Libraries/LibCore/Group.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    gr.gr_mem = nullptr;
+
+    Vector<char*> members;
+    if (group_members().size() > 0) {
+        members.ensure_capacity(group_members().size() + 1);
+        for (auto member : m_group_members) {
+            TRY(members.try_append(const_cast<char*>(member.characters())));
+        }
+        TRY(members.try_append(nullptr));
+        gr.gr_mem = const_cast<char**>(members.data());
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:102]: When this functions exits, `members` will be destroyed, so `gr.gr_mem` will point to freed memory. This is a UAF.

**Review by creator1creeper1 — CHANGES_REQUESTED** [review]

File: Userland/Utilities/groupadd.cpp [hunk 5]
```diff
@@ -2,90 +2,30 @@
...
-    if (putgrent(&g, grfile) < 0) {
-        perror("putpwent");
-        return 1;
-    }
+    args_parser.add_positional_argument(group_name, "Name of the group (groupname)", "group");
+    args_parser.parse(arguments);
 
-    fclose(grfile);
+    auto group = Core::Group { group_name, gid };
+    TRY(Core::Group::add_group(group));
```
Comment (creator1creeper1) [inline Userland/Utilities/groupadd.cpp:28]: Neat :^)

**Review by creator1creeper1 — APPROVED** [review]

**Comment by creator1creeper1** [comment]
I'm not sure what the best way to avoid these UAFs would be. We could use a function similar to `getgrnam_r()` and `getgrgid_r()`. They require the caller to provide a buffer of a certain size. This does make the interface a little more annoying, but since it's only an implementation detail (the *actual* caller will only ever be concerned with `Group`s and not with `struct group`s), I think it's a reasonable change.

File: Userland/Libraries/LibCore/Group.cpp (hunk 3 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibCore/Group.cpp:94]: I do not see an issue with that. `to_libc_group()` is a private convenience method which is only called right before we write the group entry to disk.

File: Userland/Libraries/LibCore/Group.cpp (hunk 4 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibCore/Group.cpp:109]: Thanks for noticing this, I have added a fix for this

File: Userland/Libraries/LibCore/Group.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    gr.gr_gid = group_id();
+    gr.gr_mem = nullptr;
+
+    // FIXME: A better solution would surely be not using static here?
+    // Not using a static here would result in gr.gr_mem being freed up on exit from this function.
+    static Vector<char*> members;
+    members.clear_with_capacity();
+    if (group_members().size() > 0) {
+        members.ensure_capacity(group_members().size() + 1);
+        for (auto member : m_group_members)
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:105]: Why are you using `m_group_members` in this line but `group_members()` in the lines above this?

File: Userland/Libraries/LibCore/Group.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            TRY(members.try_append(const_cast<char*>(member.characters())));
+        TRY(members.try_append(nullptr));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:107]: You could probably first use `try_ensure_capacity()` and then `unchecked_append()`, since you know the size in advance.

File: Userland/Libraries/LibCore/Group.cpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    , m_group_id(group_id)
+    , m_group_members(move(group_members))
+{
+}
+
+ErrorOr<struct group> Group::to_libc_group()
+{
+    struct group gr;
+    memset(&gr, 0, sizeof(struct group));
+    gr.gr_name = const_cast<char*>(group_name().characters());
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:94]: If it's only a private method, I guess it's okay if we force the caller to consider the lifetimes. Can you add a comment to the code to make sure that this is clear? Something along the lines of `//NOTE: *this must outlive the returned struct group`.

File: Userland/Libraries/LibCore/Group.cpp [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    : m_group_name(move(group_name))
+    , m_group_id(group_id)
+    , m_group_members(move(group_members))
+{
+}
+
+ErrorOr<struct group> Group::to_libc_group()
+{
+    struct group gr;
+    memset(&gr, 0, sizeof(struct group));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:93]: Why is this necessary? Aren't you assigning to the members of the struct anyway? Does the standard require this? I don't see anything about such a requirement on the `getgrnam()` man page (which introduces `struct group`).

File: Userland/Libraries/LibCore/Group.cpp [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    gr.gr_name = const_cast<char*>(group_name().characters());
+    gr.gr_passwd = const_cast<char*>("x");
+    gr.gr_gid = group_id();
+    gr.gr_mem = nullptr;
+
+    // FIXME: A better solution would surely be not using static here?
+    // Not using a static here would result in gr.gr_mem being freed up on exit from this function.
+    static Vector<char*> members;
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:101]: This now means that there cannot be multiple `struct group`s at the same time, because only one `gr.gr_mem` can ever be valid at the same time. While this might be a legitime design choice, you should explicitely note this to the caller as well.

**Review by creator1creeper1 — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/Group.cpp (hunk 6 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibCore/Group.cpp:106]: Changed to refer member variables instead of associated member functions

File: Userland/Libraries/LibCore/Group.cpp (hunk 9 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibCore/Group.cpp]: Agree, it is not necessary. Me being a code monkey looking at code in `grp.cpp`. I also notice now that it was actually the clean up of a used global static struct group: https://github.com/SerenityOS/serenity/blob/f92172c7c3928fa237b9059921a99d919773c63d/Userland/Libraries/LibC/grp.cpp#L50

**Comment by kennethmyhra** [comment]
Thanks for the feedback @creator1creeper1. Everything should now be addressed

File: Userland/Libraries/LibCore/Group.cpp [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    gr.gr_gid = m_group_id;
+    gr.gr_mem = nullptr;
+
+    // FIXME: A better solution would surely be not using a static here
+    // NOTE: This now means that there cannot be multiple struct groups at the same time, because only one gr.gr_mem can ever be valid at the same time.
+    // NOTE: Not using a static here would result in gr.gr_mem being freed up on exit from this function.
+    static Vector<char*> members;
+    members.clear_with_capacity();
+    if (m_group_members.size() > 0) {
+        TRY(members.try_ensure_capacity(group_members().size() + 1));
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCore/Group.cpp:105]: You missed a call to `group_members()`, just make this `m_group_members` to be consistent :^)

**Review by creator1creeper1 — CHANGES_REQUESTED** [review]
Just one last change please

File: Userland/Libraries/LibCore/Group.cpp (hunk 11 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibCore/Group.cpp]: Done:)

**Review by creator1creeper1 — APPROVED** [review]

**Review by linusg — CHANGES_REQUESTED** [review]
`unsigned int` -> `unsigned` as per https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#types

**Comment by kennethmyhra** [comment]
Whoops, fixed!

**Comment by kennethmyhra** [comment]
Noticed I had missed the initalizer for the unsigned member variable `m_id`, that's also addressed now.

File: Userland/Libraries/LibCore/Group.cpp [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+            group_id++;
+        }
+        group.set_group_id(group_id);
+    }
+
+    auto gr = TRY(group.to_libc_group());
+
+    errno = 0;
+    FILE* file = fopen("/etc/group", "a");
+    if (errno)
```
Comment (awesomekling) [inline Userland/Libraries/LibCore/Group.cpp:49]: Any reason we're not using the standard error checking condition `if (!file)` here?

File: Userland/Libraries/LibCore/Group.cpp [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        return Error::from_errno(errno);
+
+    int rc = putgrent(&gr, file);
+    if (rc < 0)
```
Comment (awesomekling) [inline Userland/Libraries/LibCore/Group.cpp:53]: Since we don't need `rc` for anything after this point, I suggest losing the local:
```cpp
if (putgrent(&gr, file) < 0)
```

File: Userland/Libraries/LibCore/Group.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+    Group() = default;
+    Group(String name, unsigned id = 0, Vector<String> members = {});
+
+    ~Group() = default;
+
+    String const& name() const { return m_name; }
+    void set_name(String const& name) { m_name = name; }
+
+    unsigned id() const { return m_id; }
```
Comment (awesomekling) [inline Userland/Libraries/LibCore/Group.h:30]: Let's use the system's `gid_t` type for GIDs everywhere.

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/Group.cpp (hunk 12 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibCore/Group.cpp]: Thought I had read the man page for `fopen()` properly. Looked it up again, and apparently not. Changed this to `if (!file)` as suggested
